File: solutions/php/all-your-base/1/AllYourBase.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

function rebase(int $fromBase, array $digits, int $toBase): array
{
    // 1. Validation des bases
    if ($fromBase < 2) {
        throw new InvalidArgumentException('input base must be >= 2');
    }

    if ($toBase < 2) {
        throw new InvalidArgumentException('output base must be >= 2');
    }

    // 2. Nettoyage des zéros inutiles au début (ex: [0, 0, 1] -> [1])
    while (count($digits) > 1 && $digits[0] === 0) {
        array_shift($digits);
    }

    // 3. Validation des chiffres et conversion de fromBase -> Base 10 (nombre decimal)
    $decimalValue = 0;
    foreach ($digits as $digit) {
        if ($digit < 0 || $digit >= $fromBase) {
            throw new InvalidArgumentException('all digits must satisfy 0 <= d < input base');
        }
        $decimalValue = $decimalValue * $fromBase + $digit;
    }

    // 4. Cas particulier : le nombre vaut 0
    if ($decimalValue === 0) {
        return [0];
    }

    // 5. Conversion de Base 10 -> toBase
    $resultDigits = [];
    while ($decimalValue > 0) {
        array_unshift($resultDigits, $decimalValue % $toBase);
        $decimalValue = intdiv($decimalValue, $toBase);
    }

    return $resultDigits;
}

File: solutions/python/all-your-base/1/all_your_base.py
def rebase(input_base, digits, output_base):
    # 1. Validation des bases
    if input_base < 2:
        raise ValueError("input base must be >= 2")
    if output_base < 2:
        raise ValueError("output base must be >= 2")

    # 2. Validation des chiffres
    for digit in digits:
        if not (0 <= digit < input_base):
            raise ValueError("all digits must satisfy 0 <= d < input base")

    # 3. Conversion de la input_base vers un entier décimal unique
    decimal_value = 0
    for digit in digits:
        decimal_value = decimal_value * input_base + digit

    # Cas particulier : le nombre vaut 0
    if decimal_value == 0:
        return [0]

    # 4. Conversion de l'entier décimal vers la output_base
    output_digits = []
    while decimal_value > 0:
        output_digits.append(decimal_value % output_base)
        decimal_value //= output_base

    return output_digits[::-1]